features/crc/mbed_crc.h

```diff
@@ -0,0 +1,223 @@
+typedef enum crc_polynomial_type {
+    CRC_7BIT_SD,
+    CRC_8BIT_CCITT,
+    CRC_16BIT_CCITT,
+    CRC_16BIT_SD,
+    CRC_16BIT_IBM,
+    CRC_32BIT
+} crc_polynomial_type_t;
+
+/** CRC width
+ *
+ * CRC value can be 8/16/32 bit
+ */
+typedef enum crc_width {
+    CRC_INVALID = 0,
+    CRC_7 = 7,
+    CRC_8 = 8,
+    CRC_16 = 16,
+    CRC_32 = 32,
+} crc_width_t;
+
+/** CRC object provides CRC generation through hardware/software
+ */
+class CRC
+{
+public:
+    typedef uint64_t crc_data_size_t;
+    typedef uint32_t crc_polynomial_t;
+    typedef uint32_t crc_size_t;
+
+    /** Lifetime of CRC object
+     */
+    virtual ~CRC() {};
+
+    /** Initialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init()
+    {
+        return 0;
+    }
+
+    /** Deinitialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t deinit()
+    {
+        return 0;
+    }
+
+    /** Get the current CRC polynomial type
+     *
+     * @return  Polynomial type
+     */
+    virtual crc_polynomial_type_t get_polynomial_type(void) const = 0;
+
+    /** Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual crc_polynomial_t get_polynomial(void) const
+    {
+        crc_polynomial_type_t polynomial = this->get_polynomial_type();
+
+        if (CRC_7BIT_SD == polynomial) {
+            return 0x9;             // x7+x3+1;
+        } else if (CRC_8BIT_CCITT == polynomial) {
+            return 0x07;            // x8+x2+x+1
+        } else if ((CRC_16BIT_SD == polynomial) || (CRC_16BIT_CCITT == polynomial)) {
+            return 0x1021;          // x16+x12+x5+1
+        } else if (CRC_16BIT_IBM == polynomial) {
+            return 0x8005;          // x16+x15+x2+1
+        } else if (CRC_32BIT == polynomial) {
+            return 0x04C11DB7;     // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
+        }
+        return 0;
+    }
+
+    /** Get the current CRC width
+     *
+     * @return  CRC width
+     */
+    virtual crc_width_t get_width(void) const
+    {
+        crc_polynomial_type_t polynomial = this->get_polynomial_type();
+
+        if (CRC_7BIT_SD == polynomial) {
+            return CRC_7;
+        } else if (CRC_8BIT_CCITT == polynomial) {
+            return CRC_8;
+        } else if ((CRC_16BIT_CCITT == polynomial) ||
+                (CRC_16BIT_SD == polynomial) ||
+                (CRC_16BIT_IBM == polynomial)) {
+            return CRC_16;
+        } else if (CRC_32BIT == polynomial) {
+            return CRC_32;
+        }
+        return CRC_INVALID;
+    }
+
+    /** Compute CRC for the data input
+     *
+     *  @param [IN]  buffer  Data bytes
+     *  @param [IN]  size  Size of data
+     *  @param [OUT] crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, crc_size_t *crc) = 0;
```

Comment:
Why are we using/calling the `crc_size_t`? It reassembles library type `ssize_t` and I would expect it to represent size. How about `crc_sum_t`?

Reply:
My preference would be to use uint32_t, but yes can have it as crc_sum_t